Bang & Olufsen add beolink grouping

mj23000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add additional media player grouping to the bang_olufsen media_player entity, using Beolink custom actions.

Beolink is not limited to Mozart devices, which means that devices that can't to be added to Home Assistant (with this integration) can be expanded to using the custom actions, which reference the device's unique JID, instead of entity_id.

Additionally extra_state_attributes show Beolink self (name and JID), peers, listeners and leader.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31880

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mj23000 committed:
Add Beolink custom services
Add support for media player grouping via beolink
Give media player entity name

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Fix progress not being set to None as Beolink listener
Revert naming changes

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Update API
simplify Beolink attributes

mj23000 committed:
Improve beolink custom services

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Fix Beolink expandable source check
Add unexpand return value
Set entity name on initialization

mj23000 committed:
Handle entity naming as intended

mj23000 committed:
Fix "null" Beolink self friendly name

joostlek reviewed (dismissed):
Please split this PR. One for bumping the library (and all changes needed to support that new version (not the changes needed for new features made possible for that version)). Then try to split the services into a separate PR if that makes sense. I haven't checked the media_player diff, but I don't believe adding services would add that much code to media_player.py

mj23000 commented:
> Please split this PR. One for bumping the library (and all changes needed to support that new version (not the changes needed for new features made possible for that version)). Then try to split the services into a separate PR if that makes sense. I haven't checked the media_player diff, but I don't believe adding services would add that much code to media_player.py

Of course. 
I've made a new PR for just the [API update](https://github.com/home-assistant/core/pull/113745). This PR depends on the API update, so I'll just keep this as a draft for now, and merge any changes that may be made when the time comes.

As for the custom services, there are 5 in total and they are all "required" for a smooth Beolink experience. These include "join", "expand", "unexpand", "leave" and "allstandby".

I could reduce this down to only the "expand" service and remove the extra_state_attributes for the Beolink session, which would retain support for media_player grouping and remove a sizable amount of code, if that would make this PR more digestable.

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Add regex service input validation
Add all_discovered to beolink_expand service
Improve beolink_expand response

mj23000 committed:
Add service icons

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Fix merge
Remove unnecessary assignment

emontnemery reviewed (commented):
There are quite a lot of small improvements going on in this PR which is not relevant for the added functionality.
Please move those changes to a separate PR, then rebase this PR once those changes have been merged.

To name a few:
- Sorting of supported features
- Improved type annotations
- coroutine functions which never yield been changed to regular functions

mj23000 committed:
Remove invalid typing
Update response typing for updated API

mj23000 committed:
Revert to old typed response dict method
Remove mypy ignore line
Fix jid possibly used before assignment

mj23000 committed:
Re add debugging logging

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Fix coroutine
Fix formatting

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Remove unnecessary update control

mj23000 commented:
@emontnemery yeah sorry about that. The improvements that you mentioned have now been merged into the dev branch through other PRs as mentioned.

emontnemery commented:
Please rebase the PR since it conflicts with the recent addition of announcements.
Also, the bang&olufsen integration is very quickly gaining new features. Please consider adding tests before we add additional features.

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Make tests pass
Fix remote leader media position bug
Improve remote leader BangOlufsenSource comparison

mj23000 committed:
Fix naming and add callback decorators

mj23000 committed:
Merge branch 'home-assistant:dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'bang_olufsen_add_beolink_grouping' of https://github.com/mj23000/core into bang_olufsen_add_beolink_grouping

mj23000 committed:
Move regex service check to variable
Suppress KeyError
Update tests

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -299,19 +402,20 @@ async def _async_update_sources(self) -> None:
         self._sources = self._audio_sources | self._video_sources
 
-        # HASS won't necessarily be running the first time this method is run
-        if self.hass.is_running:
```
How come this check is now removed?

mj23000 replied:
Yeah, that shouldn't have been removed. I'll add it again.

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -299,19 +402,20 @@ async def _async_update_sources(self) -> None:
 
     @callback
-    def _async_update_playback_metadata(self, data: PlaybackContentMetadata) -> None:
+    async def _async_update_playback_metadata(
```
Is the name of the method now misleading since it no longer just updates the image metadata?

mj23000 replied:
This is to show the correct friendly name in the extra state attribute `beolink` -> `self`

mj23000 replied:
Most of these method names refer to which WebSocket event triggers them, but yeah this should probably be changed.
Also, do you know if the `@callback` decorator can be used for async functions? It looks like it is only used 6 other places in Home Assistant and its use has been questioned here https://github.com/home-assistant/core/pull/121209#discussion_r1694788658

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -358,6 +462,135 @@ def _async_update_volume(self, data: VolumeState) -> None:
 
+    @callback
+    async def _async_update_name_and_beolink(self) -> None:
+        """Update the device friendly name."""
```
Should this happen in a separate PR, or is it somehow directly related to the grouping?

mj23000 replied:
This is to show the correct friendly name in the extra state attribute `beolink` -> `self`

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -358,6 +462,135 @@ def _async_update_volume(self, data: VolumeState) -> None:
+
+        self._remote_leader = self._playback_metadata.remote_leader
+
+        # Temp fix for mismatch in WebSocket metadata and "real" REST endpoint where the remote leader is not deleted.
```
If this is a temp fix, what is the real fix?

mj23000 replied:
I just checked and this seems to have been fixed, so I'll just remove this.

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -465,6 +701,17 @@ def source(self) -> str | None:
+
+        # Add Beolink attributes
+        if self._beolink_attribute:
+            attributes.update(self._beolink_attribute)
```
Is all this data really useful as state attributes, or should some or all of it instead be returned by a service? How do you expect users to make use of these state attributes?

mj23000 replied:
The attributes could be returned by services, but I think that would complicate overview of the whole situation, which of course is only relevant if the user has many Beolink compatible devices.

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -703,3 +950,88 @@ async def async_browse_media(
+    async def async_join_players(self, group_members: list[str]) -> None:
+        """Create a Beolink session with defined group members."""
+
+        # Use the touch to join if no entities have been defined
```
What does "the touch to join" refer to?

mj23000 replied:
Touch to join is the term used joining a Beolink Session on a Bang & Olufsen device by interacting with the product https://support.bang-olufsen.com/hc/en-us/articles/4411572885777-How-does-my-speaker-Join-an-ongoing-experience. Touch to join here refers to the command that triggers this same behavior.

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -703,3 +950,88 @@ async def async_browse_media(
+
+            # Invalid entity
+            if jid is None:
+                _LOGGER.warning("Error adding %s to group", group_member)
```
Should we raise instead of continuing on, or is that too harsh? What if one of the entity ids is not a B&O media player, should we raise then?

mj23000 replied:
Yes I think raising here is the correct choice.

emontnemery reviewed (commented):
A few comments.

Should we maybe split this PR to an initial PR which adds only the standard Home Assistant service API for creating groups and a follow-up PR which adds the custom B&O services?

mj23000 commented:
> A few comments.
> 
> Should we maybe split this PR to an initial PR which adds only the standard Home Assistant service API for creating groups and a follow-up PR which adds the custom B&O services?

Yes, I'll address some of your comments and then make a new PR for basic Home Assistant grouping.

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -97,6 +114,60 @@ async def async_setup_entry(
+
+    platform.async_register_entity_service(
+        name="beolink_join",
+        schema={vol.Optional("beolink_jid"): jid_regex},
```
The main use for the `_beolink_jid` attribute is simply to expose users to the Beolink JID of the specific product, which will then enable users to join or expand to that specific device. Not defining the JID will cause the device to join the latest Beolink group if available or any currently playing device. Additionally JIDs of all devices on the network can be found in the `beolink`->`peers` dict defined in the extra state attributes.

emontnemery replied:
So every B&O media player entity will in its state attributes have a data structure which contains the JID of all other known B&O media player?

I don't think that's great, how about:
- Let each B&O media player entity have its own JID exposed as a state attribute
- Add a service which responds with all devices found on the network
- The custom service can be called by users, and in the future also for a custom B&O configuration panel or similar

mj23000 replied:
The attributes currently look something like this:

```yaml
group_members:
  - media_player.beosound_balance_22222222
  - media_player.beosound_balance_11111111
...
beolink:
  self:
    Beosound Balance-11111111: 1111.1111111.11111111@products.bang-olufsen.com
  peers:
    Beosound Balance-11111111: 1111.1111111.11111111@products.bang-olufsen.com
    Beosound Balance-22222222: 2222.2222222.22222222@products.bang-olufsen.com
    Bathroom speaker: 3333.3333333.33333333@products.bang-olufsen.com
  leader:
    Beosound Balance-22222222: 2222.2222222.22222222@products.bang-olufsen.com
```

I don't think that it makes more sense to replace this with service calls, service calls (for peers, listeners etc.) together with the extra attributes could be useful.

joostlek replied:
Maybe I am missing something, but why don't we just target another device instead of a jid?

mj23000 replied:
Devices that aren't supported by this integration (and therefore aren't available in Home Assistant) can expanded to by using the JID.

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -97,6 +114,60 @@ async def async_setup_entry(
+        schema={
+            vol.Exclusive("all_discovered", "devices", ""): cv.boolean,
+            vol.Exclusive(
+                "beolink_jids",
```
In the extra state attributes `beolink`-> `peers`

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -133,6 +204,11 @@ def __init__(self, entry: ConfigEntry, client: MozartClient) -> None:
         self._video_sources: dict[str, str] = {}
 
+        # Beolink
+        self._beolink_sources: dict[str, bool] = {}
```
This is explained where the variable is set, is that not enough?

emontnemery replied:
That comment says:
```py
        # Some sources are not Beolink expandable. _source_change, which is used throughout the entity does not have this information.
        # Save expandable sources for Beolink services
```

It's still not clear to me what the key and values represent after reading that.
Maybe "Beolink expandable" has some clear meaning to you, it doesn't for me.

mj23000 replied:
I've improved the explanation now.

mj23000 committed:
Merge branch 'home-assistant:dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Re-add hass running check

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -133,6 +204,11 @@ def __init__(self, entry: ConfigEntry, client: MozartClient) -> None:
+        # Beolink
+        self._beolink_sources: dict[str, bool] = {}
+        self._remote_leader: BeolinkLeader | None = None
+        self._beolink_attribute: dict[str, dict] = {}
```
Yes, I'll improve typing, naming and comments.

mj23000 committed:
Improve comments, naming and type hinting

mj23000 committed:
Remove old temporary fix

mj23000 committed:
Convert logged warning to raised exception for invalid media_player
Simplify code using walrus operator

mj23000 committed:
Fix test for invalid media_player grouping

mj23000 committed:
Improve method naming

emontnemery commented on tests/components/bang_olufsen/test_media_player.py:
To make the implementation easier to understand, please add syrupy state snapshot assertions, maybe also snapshots of entity registry and device registry entries.

In particular, this would be helpful to understand the extra attributes.

mj23000 replied:
I've added syrupy assertions for entity states for the Beolink tests.

mj23000 committed:
Improve _beolink_sources explanation

mj23000 committed:
Improve _beolink_sources explanation

mj23000 commented:
I've made a smaller PR with support for Home Assistant grouping https://github.com/home-assistant/core/pull/123020

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Fix tests

mj23000 committed:
Remove service responses
Fix and add tests

mj23000 committed:
Change service to action where applicable

mj23000 committed:
Show playback progress for listeners

mj23000 committed:
Fix testing

mj23000 committed:
Remove useless initialization

emontnemery commented:
@mj23000 please don't forget to click the "Ready for review" button when you want the PR to be reviewed

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

emontnemery reviewed (commented):
A few comments. I've not reviewed the tests.

mj23000 committed:
Fix allstandby name

mj23000 committed:
Fix various casts with assertions
Fix comment placement
Fix group leader group_members rebase error
Replace entity_id method call with attribute

mj23000 commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -25,6 +25,50 @@
+      "description": "Leave a Beolink experience."
+    },
+    "beolink_allstandby": {
+      "name": "Beolink allstandby",
```
Of course. I've replaced it with "all standby".

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -350,31 +416,69 @@ def _async_update_volume(self, data: VolumeState) -> None:
+        # Update device name
+        device_registry = dr.async_get(self.hass)
+        device_registry.async_update_device(
+            device_id=cast(DeviceEntry, self.device_entry).id,
```
Yes!

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -350,31 +416,69 @@ def _async_update_volume(self, data: VolumeState) -> None:
+
+        # Add Beolink self
+        assert self.device_entry
+        assert self.device_entry.name
```
The `_async_update_beolink` is also called directly on `WebsocketNotification.BEOLINK` notifications, so I don't think the `if TYPE_CHECKING:` can be added. I've added explicit None checks ("" is also a valid name).

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -350,31 +416,69 @@ def _async_update_volume(self, data: VolumeState) -> None:
         self._remote_leader = self._playback_metadata.remote_leader
 
+        # Add Beolink listeners / leader
+
```
Yes. It's been moved back.

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -350,31 +416,69 @@ def _async_update_volume(self, data: VolumeState) -> None:
-                leader
-                if leader is not None
-                else f"leader_not_in_hass-{self._remote_leader.friendly_name}"
+                cast(str, self._get_entity_id_from_jid(self._remote_leader.jid))
```
Yeah sorry. Fixed now.

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -350,31 +416,69 @@ def _async_update_volume(self, data: VolumeState) -> None:
-            group_members.append(self.entity_id)
+            group_members.append(
+                cast(str, self._get_entity_id_from_jid(self._beolink_jid))
+            )
```
A string should always be returned as I would expect the entity to be available, but the method call is useless, I'll just replace it with `self.entity_id` as I've done below.

mj23000 commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -831,23 +958,30 @@ async def async_join_players(self, group_members: list[str]) -> None:
+        await self.async_beolink_leave()
 
-    async def _async_beolink_join(self) -> None:
+    # Custom actions:
```
Just a separator for standard Home Assistant actions and custom actions.

mj23000 commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -25,6 +25,50 @@
+      "fields": {
+        "beolink_jid": {
+          "name": "Beolink JID",
+          "description": "Manually specify Beolink JID to join."
```
JIDs are found using the beolink extra_state_attributes. 
Great idea to add a section to the services for the JID parameters. Though it seems that sections have a bug where they need to have at least one required field outside of the section to work and examples from within a section seem to be broken as well, so it hasn't been possible for me to make use of them, but I'll try to investigate further and open an issue if needed.

emontnemery replied:
> it seems that sections have a bug where they need to have at least one required field outside of the section to work and examples from within a section seem to be broken as well, so it hasn't been possible for me to make use of them

OK, that's not great. Could you share an example reproducing those two issues?

mj23000 replied:
> > it seems that sections have a bug where they need to have at least one required field outside of the section to work and examples from within a section seem to be broken as well, so it hasn't been possible for me to make use of them
> 
> OK, that's not great. Could you share an example reproducing those two issues?
> 
> Edit: Support for sections was forgotten in the action/service developer tool's YAML mode, is that what you've been testing?

The ui mode was not working either. An example of this bug is the the `beolink_join` service:
```python
    platform.async_register_entity_service(
        name="beolink_join",
        schema={vol.Optional("beolink_jid"): jid_regex},
        func="async_beolink_join",
    )
```
It has a single Optional parameter. If this parameter is added to a section:
```yaml
beolink_join:
  target:
    entity:
      integration: bang_olufsen
      domain: media_player
    device:
      integration: bang_olufsen
  fields:
    jid_options:
      collapsed: true
      fields:
        beolink_jid:
          required: false
          example: 1111.2222222.33333333@products.bang-olufsen.com
          selector:
            text:
```
Then the service can still be triggered with or without defining the `beolink_jid` parameter in YAML, but there is no section and no `beolink_jid` parameter in the developer tools' ui mode, only an `Action data` raw YAML field. There is also a "Parameters only available in YAML mode" message that says the section (jid_options) is only available in YAML:
<img width="1233" alt="Screenshot 2024-09-26 at 12 27 03" src="https://github.com/user-attachments/assets/fda5ed43-e13b-4ea5-a929-e6d4c3c01260">

Adding a parameter (beolink_jid_) outside of the section makes the section visible:
<img width="1233" alt="Screenshot 2024-09-26 at 12 34 37" src="https://github.com/user-attachments/assets/e01576c2-b828-45f2-8854-7a2ff1f451fd">

But the section breaks as soon as I enter anything into it and example values only fill in outside of the section:
<img width="1233" alt="Screenshot 2024-09-26 at 12 37 35" src="https://github.com/user-attachments/assets/639024f4-9dd2-4470-a16b-3ad7396a308a">

emontnemery replied:
There were previous bugs related to collapsible bugs when in UI mode, also the bug when in YAML mode has been fixed by https://github.com/home-assistant/frontend/pull/22096.

With the snippet above, UI mode displays like this:
![image](https://github.com/user-attachments/assets/bfeceece-ae99-4221-96a5-325ec167b9ed)

And YAML mode like this:
![image](https://github.com/user-attachments/assets/f226d399-c938-45d8-9c81-bdb5130b180c)

mj23000 replied:
Nice! I've added sections to the  Beolink services that use JIDs directly.

mj23000 replied:
I just tested it again and it seems to display and act as expected for developer tools and the automation editor.

emontnemery replied:
Thanks for confirming 👍

mj23000 committed:
Add syrupy snapshots for Beolink tests, checking entity states
Use test JIDs 3 and 4 instead of 2 and 3 to avoid invalid attributes in testing

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Add sections for fields using Beolink JIDs directly

mj23000 committed:
Fix typo

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_beolink_grouping

emontnemery commented on homeassistant/components/bang_olufsen/media_player.py:
```diff
@@ -116,6 +124,110 @@ async def async_setup_entry(
+        func="async_beolink_allstandby",
+    )
+
+    # Register actions.
+    platform = async_get_current_platform()
+
+    jid_regex = vol.Match(
+        r"(^\d{4})[.](\d{7})[.](\d{8})(@products\.bang-olufsen\.com)$"
+    )
+
+    platform.async_register_entity_service(
+        name="beolink_join",
+        schema={vol.Optional("beolink_jid"): jid_regex},
```
I think there's another rebase mistake here, this seems to be a duplicate of what's at line 127?

mj23000 replied:
Sorry about that

emontnemery commented on homeassistant/components/bang_olufsen/services.yaml:
Please sort the top-level keys alphabetically:
```yaml
beolink_allstandby:
...
beolink_expand:
...
etc
```

emontnemery commented on homeassistant/components/bang_olufsen/strings.json:
```diff
@@ -25,6 +29,68 @@
     }
   },
+  "services": {
+    "beolink_join": {
```
please sort 🔡

emontnemery reviewed (changes requested):
I think there's another rebase mistake to clear up

mj23000 committed:
FIx rebase mistake

mj23000 committed:
Sort actions alphabetically

emontnemery reviewed (approved):
Thanks, @mj23000 👍
